Closes #31 wrapper functions for waist/hip and waist/height ratio

yurovska opened the pull request:
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral family codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral<ext> (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the Reference page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

yurovska commented:
Please don’t judge too harshly, this is my first Pull Request ever.

yurovska commented:
@manciniedoardo Need help with `{units}` package. Checks failed due to its absence within the container.

manciniedoardo commented:
> @manciniedoardo Need help with `{units}` package. Checks failed due to its absence within the container.

@AndersAskeland can you look as well pls?

AndersAskeland commented:
@yurovska It probably fails as the units package depends on a system dependency of udunits2. 

This library needs to be manually installed in Unix (test suite that fails is running this) and MacOS. On Windows it is included in RTools as "bin\udunits2.exe".

As such, we unfortunately cannot use units as a dependency. I would actually also argue that even if it did not have a system dependency, we should try refraining from importing it, as we should try to limit our dependencies. I have not looked at your code yet, but I am guessing that you use the package to convert between units (m/cm/inches) in cases where waist is collected in a different unit to hip.

I think the concept of allowing different units is important, as one cannot always control what the collected/SDTM units are. I would actually argue that this functionality should be included in admiral base at some point

As for alternative approaches than using the units package, I can personally think of two potential solutions. Either we do manual unit conversion inside the wrapper functions (probably messy to have all potential units conversions inside `derive_param_ratio()`) or we create some internal data that controls unit conversions and do this in `derive_param_ratio()`. In either case, I would base the potential unit conversions on the permissible units defined in the CDISC unit codelist.

I think that we might need to discuss this before implementation though, and maybe create a separate issue and pull request.

yurovska commented:
> This library must be manually installed in Unix (test suite that fails is running this) and MacOS. On Windows it is included in RTools as "bin\udunits2.exe".

I'm on MacOS and just did `install.packages("units")` with no additional steps. All good.

> As such, we unfortunately cannot use the `units` package as a dependency. I would actually also argue that even if it did not have a system dependency, we should try refraining from using it, as we should try to limit our dependencies.

It's a pity. `{units}` really works like a magic. I simply set the units from `AVALU` (or `admiral::extract_unit(PARAM)`) to the context of `AVAL` in a preliminary step and the conversion performs automatically when `AVAL` is used in the formula, so that the values of two input parameters are unified prior to Ratio derivation. Then I remove units from `AVAL` context before returning the result. That easy.

yurovska commented:
@AndersAskeland Just tested it also within Roche environment (Ubuntu 22.04 Jammy). No issues.

Do you really think it might be missing in any environment except GitHub CI/CD?

```R
> renv::install("units", repos = "https://packages.roche.com/Validated-4.3/latest")
# Downloading packages -------------------------------------------------------
- Downloading units from CRAN ...               OK [file is up to date]
Successfully downloaded 1 package in 1.3 seconds.

The following package(s) will be installed:
- units [0.8-5]
These packages will be installed into "~/R/x86_64-pc-linux-gnu-library/4.3".

Do you want to proceed? [Y/n]: Y

# Installing packages --------------------------------------------------------
- Installing units ...                          OK [installed binary and cached in 1.9s]
Successfully installed 1 package in 2 seconds.
> library(units)
udunits database from /usr/share/xml/udunits/udunits2.xml
> val_cm <- set_units(1:10, cm)
> val_cm
Units: [cm]
 [1]  1  2  3  4  5  6  7  8  9 10
> val_mm <- set_units(1:10, mm)
> val_mm
Units: [mm]
 [1]  1  2  3  4  5  6  7  8  9 10
> val_cm - val_mm
Units: [cm]
 [1] 0.9 1.8 2.7 3.6 4.5 5.4 6.3 7.2 8.1 9.0
```

AndersAskeland commented:
> @AndersAskeland Just tested it also within Roche environment (Ubuntu 22.04 Jammy). No issues.
> 
> Do you really think `udunits2` might be missing in any environment except GitHub CI/CD?
> 

I would guess most compute environment have it installed. However, I still feel that we cannot expect all users to have this library or possibility of building it from source. 

I noticed that in our compute environment, when first installing `units`, udunits2 is built from source using the g++/GCC compiler.

So while it might usually work, we would put a requirement on users to have a working C compiler, which is not always the case.

For example for Mac, I think the brew process (or MacOS developer tools) installs the required compilers, however, if you just install R with nothing else, I am unsure if it will work. There might also be a difference on ARM Macs.

If we however take a step back and look at what it is used for, is it only for converting meters/mm/cm/inches? Do you use it for more?

yurovska commented:
@AndersAskeland Re-written without use of {units} package. However, only units of length are supported now.

AndersAskeland commented:
@yurovska Nice! I will start looking at the pull request. It might take some time, as this is the first time reviewing functions in Admiral and I need to figure out the general layout and stuff :)

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' Adds a Parameter for Waist to Hip Ratio
+#'
+#' @description Adds a record for Waist to Hip Ratio using Waist Circumference and Hip Circumference
+#' each by group (e.g., subject and visit) where the source parameters are available.
```
```suggestion
#' for each by group (e.g., subject and visit) where the source parameters are available.
```

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' Adds a Parameter for Waist to Height Ratio
+#'
+#' @description Adds a record for Waist to Height Ratio using Waist Circumference and Height
+#' each by group (e.g., subject and visit) where the source parameters are available.
```
```suggestion
#' for each by group (e.g., subject and visit) where the source parameters are available.
```

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' @param wstcir_code Waist Circumference parameter code
+#'
+#'   The observations where \code{PARAMCD} equals the specified value are considered
+#'   as the Waist Circumference
```
```suggestion
#'   as the Waist Circumference.
```

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' @param height_code Height parameter code
+#'
+#'   The observations where \code{PARAMCD} equals the specified value are considered
+#'   as the Height. It is expected that Height is measured in cm
```
```suggestion
#'   as the Height. It is expected that Height is measured in cm.
```

manciniedoardo replied:
But can you have height in other units (m)? Can we add flexibility here?

yurovska replied:
Thanks for your comment. This indeed became obsolete after unit conversion on the fly has been implemented. Will remove it in the following commit.

yurovska replied:
Done

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' @param wstcir_code Waist Circumference parameter code
+#'
+#'   The observations where \code{PARAMCD} equals the specified value are considered
+#'   as the Waist Circumference
```
```suggestion
#'   as the Waist Circumference.
```

manciniedoardo commented on R/derive_bds_params.R:
```diff
@@ -0,0 +1,250 @@
+#' Adds a Ratio Parameter Computed from the Analysis Value of Other Parameters
```
If this function is a utility for the other two, in my opinion it should not have its own file - it should instead be at the top of the other file.

yurovska replied:
Negotiable but I agree to move it.

yurovska replied:
Done. Everything is in a single file now.

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#'
+#' @export
+#'
+#' @seealso \code{\link[admiral:derive_param_computed]{admiral::derive_param_computed()}}
```
add the other ratio function here too

yurovska replied:
Good point. Thanks.

yurovska replied:
All functions from the same family are actually added there automatically.

![image](https://github.com/user-attachments/assets/f5f25867-12fc-4fca-bbc9-cacb6023074a)

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#'
+#' @export
+#'
+#' @seealso \code{\link[admiral:derive_param_computed]{admiral::derive_param_computed()}}
```
add the other ratio function here too

yurovska replied:
Good point. Thanks.

yurovska replied:
Same as above

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' @seealso \code{\link[admiral:derive_param_computed]{admiral::derive_param_computed()}}
+#'
+#' @examples
+#'
```
remove empty line
```suggestion
```

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#' @seealso \code{\link[admiral:derive_param_computed]{admiral::derive_param_computed()}}
+#'
+#' @examples
+#'
```
remove empty line 
```suggestion
```

manciniedoardo commented on R/derive_bds_params.R:
```diff
@@ -0,0 +1,250 @@
+#'
+#' @rdname unit-conversion
+#' @keywords internal
+get_conv_factors_all <- function() {
```
where do these conversion factors come from? Add reference please. Also, wondering what happens if i want more precision?

manciniedoardo replied:
Ok thanks, please reference and add this explanation to the function documentation.

yurovska replied:
Added a note in the documentation.

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#'   ),
+#'   constant_by_vars = exprs(USUBJID),
+#'   get_unit_expr = AVALU
+#' )
```
Add example using constant_by_vars and filter arguments

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,317 @@
+#'     PARAM = "Waist to Hip Ratio"
+#'   ),
+#'   get_unit_expr = admiral::extract_unit(PARAM)
+#' )
```
add example using constant_by_vars and filter params

yurovska replied:
`derive_param_waisthip` doesn't have `constant_by_vars` argument as it makes no sense to measure Waist and/or Hip Circumference only once at baseline.

manciniedoardo reviewed (changes requested):
Thanks @yurovska! I left some general comments; @AndersAskeland will do a technical review.

Additionally:

- please take this opportunity to remove `hello_admiral()` and related tests/documentation as we now have a real function.
- please ensure the functions appear on the reference page and take the opportunity to tidy up the reference page if you can - happy to help on this front if needed.
- add tests for the ratio function too, otherwise the test coverage for the package will be low.

AndersAskeland commented:
> * I actually tried to add unit tests for `derive_param_ratio` but R CMD check failed because `testthat` was not able to find a function, which is not exported. Perhaps, internal functions do not require testing since it's covered by unit tests of the exported functions that use the internal ones, so it won't affect the overall test coverage rate. Please let me know if you think otherwise.

I would still think we need to do something for non-exported functions, especially on derive_var_ratio, as it is somewhat advanced. I also believe that covr by default include non-exported functions. How are you running the R CMD check? Is it working when running `devtools::check()`?

yurovska commented:
> > * I actually tried to add unit tests for `derive_param_ratio` but R CMD check failed because `testthat` was not able to find a function, which is not exported. Perhaps, internal functions do not require testing since it's covered by unit tests of the exported functions that use the internal ones, so it won't affect the overall test coverage rate. Please let me know if you think otherwise.
> 
> I would still think we need to do something for non-exported functions, especially on derive_var_ratio, as it is somewhat advanced. I also believe that covr by default include non-exported functions. How are you running the R CMD check? Is it working when running `devtools::check()`?

Hm, I tried it out again and it now works fine. Probably, I'm confusing the unit tests with `@examples` section where examples were not able to run for internal functions. Anyway, I've now managed to add a couple of tests for `derive_param_ratio`.

manciniedoardo commented:
> > > * I actually tried to add unit tests for `derive_param_ratio` but R CMD check failed because `testthat` was not able to find a function, which is not exported. Perhaps, internal functions do not require testing since it's covered by unit tests of the exported functions that use the internal ones, so it won't affect the overall test coverage rate. Please let me know if you think otherwise.
> > 
> > 
> > I would still think we need to do something for non-exported functions, especially on derive_var_ratio, as it is somewhat advanced. I also believe that covr by default include non-exported functions. How are you running the R CMD check? Is it working when running `devtools::check()`?
> 
> Hm, I tried it out again and it now works fine. Probably, I'm confusing the unit tests with `@examples` section where examples were not able to run for internal functions. Anyway, I've now managed to add a couple of tests for `derive_param_ratio`.

Glad it's fixed! Yeah any function in the package should have unit tests.

AndersAskeland commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+
+    # Check if units are the same within each of two parameters
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
+        filter(PARAMCD == param) %>%
+        pull(`_unit`) %>%
+        unique()
+
+      if (length(units_found) != 1L) {
+        cli_abort(
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
+
+      param_units[[param]] <- units_found
+    }
```
```suggestion
units_found <- dataset %>%
    filter(PARAMCD %in% c(numerator_code, denominator_code)) %>%
    group_by(PARAMCD) %>%
    mutate(`_unit` = !!get_unit_expr) |> 
    summarise(`_distinct_units` = n_distinct(`_unit`))
    
    # Check / fail
    if(any(units_found$`_distinct_units`) > 1) 
    {
        cli_abort(
          "Multiple units {.val {units_found}} found for {.val {param}}. # Need to change this a bit I guess
          Please review and update the units."
        )
      }
```

yurovska replied:
What's the rational for this change?

AndersAskeland replied:
Loops are difficult to debug and maintain if anything ever goes wrong. I would say that loops should only be used if we want some side effects (i.e. saving a file), we need to extract names to use somewhere else, or we don't know the output length (i.e. when there is some randomness somewhere). 

In this case, we will always have two parameters where we want to figure out if the unit is distinct within each param. I would argue that it is much easier to decode that functionality the non-looped code.

There are perhaps also some performance benefits to not looping, however, they would probably be tiny in this example.

However, this might be moot, as I think that we should do this functionality in a assertion. Either our own, or from admiraldev.

yurovska replied:
No loops are there anymore.

yurovska commented on R/derive_bds_params.R:
```diff
@@ -0,0 +1,250 @@
+#' Adds a Ratio Parameter Computed from the Analysis Value of Other Parameters
```
I would keep the unit conversion part in the advs file for now as the functions, although written in a more or less universal way with scalability in mind, are still too limited for a specific purpose (only length conversion and only for ratio derivation). We can move it to `utils.R`, once it becomes of help somewhere else.

AndersAskeland replied:
Agree!

yurovska commented on R/derive_bds_params.R:
```diff
@@ -0,0 +1,250 @@
+#'
+#'   *Permitted Values:* A variable of the input dataset or a function call
+#'
+#' @inheritParams admiral::derive_param_bmi
```
The reason is that `derive_param_computed` in `@param set_values_to` provides an example of ECG parameter derivation with a formula, which not relevant here. Although, I agree to replace the inheritance from BMI with a generic `derive_param_computed` and will just put my own `@param set_values_to`.

yurovska replied:
Updated

yurovska commented on R/derive_bds_params.R:
```diff
@@ -0,0 +1,250 @@
+          "Multiple units {.val {units_found}} found for {.val {param}}.
+          Please review and update the units."
+        )
+      }
```
> Thinking more about it, I noticed that this is a based on the code in `admiraldev::assert_unit()`, without the checks on required units. Perhaps, we should instead try to get it changed in `admiraldev` to firstly not have a required argument for units, and secondly add the possibility of having several units.

Some time ago I already requested a change of `admiraldev::assert_unit()` to add the possibility of having several units in `required_unit` argument and even proposed my implementation.

Discussion - https://github.com/pharmaverse/admiral/issues/2513
Feature Request - https://github.com/pharmaverse/admiraldev/issues/468

I've just also put a comment there about making `required_unit` optional.

AndersAskeland replied:
Amazing! I propose that while waiting, we should move this check to a non-exported function. Then we could add the check to the very top of the function (in alignment with admiral developer recommendation), and create a issue (maybe a "good first issue") for removing it when its available in admiraldev.

yurovska replied:
I've added the updated version of `assert_unit()` allowing multiple values in `required_unit` argument in `assertions.R` and marked it as deprecated, while waiting for https://github.com/pharmaverse/admiraldev/issues/468. We'll remove it once it's implemented in `admiraldev`. Please open an issue for that when the review is completed.

AndersAskeland replied:
Issue #47 is ready.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+  if (constant_denominator) {
+    constant_parameters <- c(constant_parameters, denominator_code)
+
+    parameters <- parameters %>%
+      .[!. == denominator_code] %>%
+      ifelse(length(.) == 0, NULL, .)
+  }
```
Thanks for the suggested change. Agree to update.

> See my other suggestion about changing how we handle constants. This could perhaps eliminate the need for this code.

This code is needed here because `derive_param_waisthgt` doesn't have `constant_parameters` argument at all where I omitted it on purpose as only HEIGHT can be constant (in line with `admiral::derive_param_bmi`). So we either identify constant parameters here or in every single wrapper separately to build `constant_parameters` argument for `derive_param_computed`. I'd better keep it in `derive_param_ratio`.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+  if (constant_numerator) {
+    constant_parameters <- c(constant_parameters, numerator_code)
+
+    parameters <- parameters %>%
+      .[!. == numerator_code]
```
Agree, thanks.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+#'
+#'   *Permitted Values:* character value
+#'
+#' @param constant_numerator Is numerator parameter constant?
+#'
+#'   It is expected that the parameter code (PARAMCD) specified in \code{numerator_code}
+#'   which is required to derive the new parameter is measured only once. For example,
+#'   if Height to Weight Ratio should be derived and height is measured only once
+#'   while Weight is measured at each visit. Height could be specified in the
+#'   \code{numerator_code} argument and \code{constant_numerator} is to be set to \code{TRUE}.
+#'
+#'   *Permitted Values:* logical scalar
+#'
```
Yeah, I just added a possibility to have either numerator or denominator being constant with other potential ratios in mind where, for example, Height could be a numerator, like Height to Weight Ratio, or whatever else.

Good point regarding setting both to TRUE. I agree that only one of two parameters can be constant. I'll add a check and a corresponding error message.

yurovska replied:
Added a check to allow only one of two parameters to be constant.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+        param_units[[numerator_code]]
+      )
+
+      if (!is.na(conv_factor)) {
+        ratio_formula <- expr(
+          !!sym(paste0("AVAL.", numerator_code)) / (
+            !!sym(paste0("AVAL.", denominator_code)) * !!conv_factor
+          )
+        )
```
Unit conversion must be optional because there might be other ratios where units of two input parameters are not convertible, e.g. when if I derived Height (cm) to Weight (kg), I would not expect any conversion factor between cm and kg. The result is just X cm/kg. Why should it fail in this case?

AndersAskeland replied:
Good point!

I however still mean that we need a way to fail if there are unsupported units in the `waisthip` and `waistheight` functions. This is now the case, when you have implemented the unit assertion function. Feel free to resolve this conversation.

yurovska replied:
See my replies in other conversations.

`get_unit_expr` is no longer optional in the `waisthip` and `waistheight` and the units are asserted with the list of supported ones.

AndersAskeland replied:
Sorry, my formulation was confusing. I agree, this appears to be handled now.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+                                  set_values_to = exprs(PARAMCD = "WAISTHGT"),
+                                  filter = NULL,
+                                  constant_by_vars = NULL,
+                                  get_unit_expr = NULL) {
```
I did it in the opposite way. If the unit conversion is applied, there will be an info message in the log about that.

Once `admiraldev::assert_unit()` is updated  to allow checking for multiple units, I would make `get_unit_expr` a required argument the waisthip and waisthgt functions and assert it with CDISC synonyms for length only.

yurovska replied:
Units are now asserted with CDISC synonyms for length, i.e. only cm, m, mm, in and ft are acceptable, so that no strange ratio is possible anymore.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+
+    for (param in c(numerator_code, denominator_code)) {
+      units_found <- dataset %>%
+        mutate(`_unit` = !!get_unit_expr) %>%
```
Not aware of this kind of guidelines, it's just based on `admiraldev::assert_unit()`

yurovska replied:
I think this is when you derive a temp var in the dataset that you're going to return. Here it's not applicable because the var created in an intermediate dataset. Although, I agree to replace with `tmp_unit`.

yurovska replied:
Done

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+
+#' @description `get_conv_factors_all()` returns all conversion factors supported.
+#'
+#' **Note:** Conversion factor for inch comes from the NCI Thesaurus
```
Updated to

`**Note:** Conversion factors comes from unit definitions as per the NCI Thesaurus`

AndersAskeland replied:
I would actually prefer to write the actual conversions/factors, even though it will be verbose. I.e. "A meter is defined as 100cm". That would make it easier for users to quickly check that the unit conversions make sense. I also think we should refer to the CDISC definitions and not NCI, even though CDISC just copy/paste the NCI definition.

yurovska replied:
Added

![image](https://github.com/user-attachments/assets/85d4f2c9-5eee-4c64-ad6f-760f8b7da7f9)

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,600 @@
+  # Get all conversion factors supported
+  conv_factors_all <- get_conv_factors_all()
+
+  # Return conversion factor if units are supported and convertible
+  for (unit_category in names(conv_factors_all)) {
+    if (all(c(from_unit, to_unit) %in% names(conv_factors_all[[unit_category]]))) {
+      return(
+        conv_factors_all[[unit_category]][[from_unit]] /
+          conv_factors_all[[unit_category]][[to_unit]]
+      )
+    }
+  }
```
I think we should keep it silent for other potential ratios where input parameters are not supposed to be convertible, like Height (cm) to Weight (kg) or so. For `derive_param_waisthip` and `derive_param_waisthgt` this is no longer a problem since I've updated them to assert units using `assert_unit()` with only those, which are supported.

yurovska commented:
@AndersAskeland @manciniedoardo Thanks again for your valuable review. I think I'm done addressing all your comments and the PR is ready for the next (or maybe final) round. I haven't resolved some conversations, please do so if you agree with the update made.

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,632 @@
+#' Adds a Parameter for Waist to Hip Ratio
+#'
+#' @description Adds a record for Waist to Hip Ratio using Waist Circumference and Hip Circumference
```
Minor: align language
```suggestion
#' @description Adds a parameter for Waist to Hip Ratio using Waist Circumference and Hip Circumference
```

manciniedoardo commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,632 @@
+
+#' Adds a Parameter for Waist to Height Ratio
+#'
+#' @description Adds a record for Waist to Height Ratio using Waist Circumference and Height
```
Minor:
```suggestion
#' @description Adds a parameter for Waist to Height Ratio using Waist Circumference and Height
```

manciniedoardo reviewed (approved):
LGTM - a couple of minor comments to implement but then I am happy from a documentation and user side. @AndersAskeland and @pharmaverse/admiralmetabolic pls add your final feedback and any technical by end of week so we can then merge.

yurovska commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,632 @@
+#' @name unit-conversion
+#' @keywords internal
+NULL
+#> NULL
```
This is from `roxygen2` documentation

https://roxygen2.r-lib.org/articles/reuse.html#rdname

AndersAskeland commented:
Sorry for my late review, I was suddenly required to travel out of office for most of the week, but will quickly go trough it tomorrow :) I have been thinking about one thing, which I think we shouldn’t necessarily change now, but could alter in the future depending on your thoughts on it.

My thoughts have been about the design/implementation of the actual conversion, and weather that should be something associated to the actual weight height / waist hip ratio, or should be more general and associated with the ratio function (which it is now). We had a short discussion about me wanting the conversion function to fail if conversion was not supported, but you (rightly) pointed out that there are ratios that we should not convert (for instance kg/m).

I still agree that this is something we should cater to . However, I am thinking about cases where you might be interested in a ratio with convertible units. For instance a meter / centimetre ratio. I realise that this is very uncommon and strange, but my thoughts are then, who are we to decide which ratios to allow? 

Hence, I think we should at some point discuss if we should keep the ratio function extremely generic and move everthing associated to conversion to the wrapper function. In some sense I think this makes more sense, because, we would not un necessarily call a convert function when not needed, and we could be more consistent with the return from the unit conversion function. I am not a big fan of a function being able to return NAs.

But… For time sakes, and to allow us to implement this before next week, I think we can discuss this later as it would not actually impede the functions to change the implementation.

yurovska commented:
> Sorry for my late review, I was suddenly required to travel out of office for most of the week. I will quickly go through it tomorrow :) I have been thinking about one thing though, which I think we shouldn’t necessarily change now, but could alter in the future depending on your thoughts on it.
> 
> My thoughts have been about the design/implementation of the actual conversion, and weather that should be something associated to the actual weight height / waist hip ratio, or should be more general and associated with the ratio function (which it is now). We had a short discussion about me wanting the conversion function to fail if conversion was not supported, but you (rightly) pointed out that there are ratios that we should not convert (for instance kg/m).
> 
> I still agree that this is something we should cater to . However, I am thinking about cases where you might be interested in a ratio with convertible units. For instance a meter / centimetre ratio. I realise that this is very uncommon and strange, but my thoughts are then, who are we to decide which ratios to allow?
> 
> Hence, I think we should at some point discuss if we should keep the ratio function extremely generic and move everthing associated to conversion to the wrapper function. In some sense I think this makes more sense, because, we would not un necessarily call a convert function when not needed, and we could be more consistent with the return from the unit conversion function. I am not a big fan of a function being able to return NAs.
> 
> But… For time sakes, and to allow us to implement this before next week, I think we can discuss this later as it would not actually impede the functions to change the implementation.

Thanks in advance for finding time to finalize the review.

I got your point and I think I have an idea. What if we add a sort of `unit_conversion` argument (TRUE/FALSE) in `derive_param_ratio` that would be disabled by default, so that a developer of a wrapper function decides whether to call it with `unit_conversion = TRUE` if applicable for a particular ratio parameter? With this approach I'd agree to throw an error if conversion factor is not found instead of returning NA.

AndersAskeland commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,633 @@
+#'   The observations where `PARAMCD` equals the specified value are considered
+#'   as the Waist Circumference.
+#'
+#'   *Permitted Values:* character value
```
This is probably more applicable to admiral, so I would not change it here, but should this not be "character string" or just "character"?

yurovska replied:
It's consistent with other admiral functions.

AndersAskeland commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,633 @@
+#'   Otherwise it will only be calculated at visits with both Height and Waist Circumference
+#'   collected.
+#'
+#'   *Permitted Values*: list of variables created by `exprs()`, e.g. `exprs(USUBJID, VISIT)`
```
I am assuming this comes from admiral? Would it not in reality be list of symbols? Or maybe add something indicating we use data masking. Here is an example from `ggplot2::aes()`

![image](https://github.com/user-attachments/assets/47253dfa-d396-4136-8c21-955952b735cf)

yurovska replied:
It's consistent with other admiral functions.

AndersAskeland commented on R/derive_advs_params.R:
```diff
@@ -0,0 +1,633 @@
+#'
+#'
+#' @return The input dataset with the new parameter added. Note, a variable will only
+#'         be populated in the new parameter rows if it is specified in `by_vars`.
```
What does "Note, a variable will only be populated in the new parameter rows if it is specified in `by_vars`."

What does this mean? I did not think that `by_vars` had anything to do with `set_values_to`?

yurovska replied:
This was also taken from admiral. It means that if, for instance, you have also `ACTARM` variable that you wanna keep for derived parameters, you have to specify it in `by_var` as well. Otherwise it won't get populated for new records.

```R
advs <- tribble(
  ~USUBJID, ~ACTARM, ~PARAMCD, ~PARAM, ~AVAL, ~AVALU, ~VISIT,
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   125, "cm", "SCREENING",
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   124, "cm", "WEEK 2",
  "01-101-1001", "ARM 1", "HIPCIR", "Hip Circumference (cm)",   123, "cm", "WEEK 3",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 43.31, "in", "SCREENING",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 42.52, "in", "WEEK 2",
  "01-101-1001", "ARM 1", "WSTCIR", "Waist Circumference (in)", 42.13, "in", "WEEK 3",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   135, "cm", "SCREENING",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   133, "cm", "WEEK 2",
  "01-101-1002", "ARM 2", "HIPCIR", "Hip Circumference (cm)",   132, "cm", "WEEK 3",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 47.24, "in", "SCREENING",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 46.46, "in", "WEEK 2",
  "01-101-1002", "ARM 2", "WSTCIR", "Waist Circumference (in)", 46.06, "in", "WEEK 3"
)

derive_param_waisthip(
  advs,
  by_vars = exprs(USUBJID, VISIT),   # No ACTARM here
  wstcir_code = "WSTCIR",
  hipcir_code = "HIPCIR",
  set_values_to = exprs(
    PARAMCD = "WAISTHIP",
    PARAM = "Waist to Hip Ratio"
  ),
  get_unit_expr = admiral::extract_unit(PARAM)
)
```

```
   USUBJID     ACTARM PARAMCD  PARAM                       AVAL AVALU VISIT    
   <chr>       <chr>  <chr>    <chr>                      <dbl> <chr> <chr>    
 1 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   125     cm    SCREENING
 2 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   124     cm    WEEK 2   
 3 01-101-1001 ARM 1  HIPCIR   Hip Circumference (cm)   123     cm    WEEK 3   
 4 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  43.3   in    SCREENING
 5 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  42.5   in    WEEK 2   
 6 01-101-1001 ARM 1  WSTCIR   Waist Circumference (in)  42.1   in    WEEK 3   
 7 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   135     cm    SCREENING
 8 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   133     cm    WEEK 2   
 9 01-101-1002 ARM 2  HIPCIR   Hip Circumference (cm)   132     cm    WEEK 3   
10 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  47.2   in    SCREENING
11 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  46.5   in    WEEK 2   
12 01-101-1002 ARM 2  WSTCIR   Waist Circumference (in)  46.1   in    WEEK 3   
13 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.880 NA    SCREENING
14 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.871 NA    WEEK 2   
15 01-101-1001 NA     WAISTHIP Waist-to-Hip Ratio         0.870 NA    WEEK 3   
16 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.889 NA    SCREENING
17 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.887 NA    WEEK 2   
18 01-101-1002 NA     WAISTHIP Waist-to-Hip Ratio         0.886 NA    WEEK 3   
```

AndersAskeland replied:
I see. I would argue that it could be made clearer, but I agree that we should be consistent with Admiral.

AndersAskeland reviewed (approved):
Amazing work @yurovska ! 

I think it is ready to merge now!

I have a couple of questions/comments, but they are more some general considerations that I don't think we should do anything about from our side :)

AndersAskeland commented:
> > Sorry for my late review, I was suddenly required to travel out of office for most of the week. I will quickly go through it tomorrow :) I have been thinking about one thing though, which I think we shouldn’t necessarily change now, but could alter in the future depending on your thoughts on it.
> > My thoughts have been about the design/implementation of the actual conversion, and weather that should be something associated to the actual weight height / waist hip ratio, or should be more general and associated with the ratio function (which it is now). We had a short discussion about me wanting the conversion function to fail if conversion was not supported, but you (rightly) pointed out that there are ratios that we should not convert (for instance kg/m).
> > I still agree that this is something we should cater to . However, I am thinking about cases where you might be interested in a ratio with convertible units. For instance a meter / centimetre ratio. I realise that this is very uncommon and strange, but my thoughts are then, who are we to decide which ratios to allow?
> > Hence, I think we should at some point discuss if we should keep the ratio function extremely generic and move everthing associated to conversion to the wrapper function. In some sense I think this makes more sense, because, we would not un necessarily call a convert function when not needed, and we could be more consistent with the return from the unit conversion function. I am not a big fan of a function being able to return NAs.
> > But… For time sakes, and to allow us to implement this before next week, I think we can discuss this later as it would not actually impede the functions to change the implementation.
> 
> Thanks in advance for finding time to finalize the review.
> 
> I got your point and I think I have an idea. What if we add a sort of `unit_conversion` argument (TRUE/FALSE) in `derive_param_ratio` that would be disabled by default, so that a developer of a wrapper function decides whether to call it with `unit_conversion = TRUE` if applicable for a particular ratio parameter? With this approach I'd agree to throw an error if conversion factor is not found instead of returning NA.

That is a good idea! That solution also has the benefit of us not needing to repeat ourselves in the "specific" functions (by needing to have code for the conversion function in both function).

yurovska commented on R/assertions.R:
```diff
@@ -0,0 +1,63 @@
+#' Asserts That a Parameter is Provided in One of the Expected Units
```
It's already been implemented there and I've reviewed it. Just waiting for PR to be merged.

https://github.com/pharmaverse/admiraldev/pull/471

yurovska commented:
> That is a good idea! That solution also has the benefit of us not needing to repeat ourselves in the "specific" functions (by needing to have code for the conversion function in both function).

I'll keep that in mind and implement as part of moving from the internal `assert_unit()` to the admiraldev's one.

yurovska commented:
> Amazing work @yurovska !
> 
> I think it is ready to merge now!
> 
> I have a couple of questions/comments, but they are more some general considerations that I don't think we should do anything about from our side :)

Thanks again for your review. Questions answered. Conversations resolved. We're good to go merging.
